### PR TITLE
feat: apply descriptor exclude_types to env mat stat

### PR DESCRIPTION
**Pull request by njzjz**

### CODE REVIEW

File: deepmd/pt/utils/env_mat_stat.py [hunk 1]
```diff
@@ -146,6 +146,9 @@ def iter(
                 radial_only,
                 protection=self.descriptor.env_protection,
             )
+            # apply excluded_types
+            exclude_mask = self.descriptor.emask(nlist, extended_atype)
+            env_mat *= exclude_mask.unsqueeze(-1)
```
Comment (wanghan-iapcm) [inline deepmd/pt/utils/env_mat_stat.py:151]: may be a better option to use the same mechanism as the atomic model 
https://github.com/deepmodeling/deepmd-kit/blob/cc30f275f34fd999658025446703289c4718aa31/deepmd/pt/model/atomic_model/dp_atomic_model.py#L209-L222

**Review by wanghan-iapcm — COMMENTED** [review]
I agree with this PR, but please not that would deprecate the exclude type in the future.

File: deepmd/tf/descriptor/se_r.py [hunk 2]
```diff
@@ -196,6 +196,18 @@ def __init__(
                 rcut_smth=self.rcut_smth,
                 sel=self.sel_r,
             )
+            if len(self.exclude_types):
+                # exclude types applied to data stat
+                mask = self.build_type_exclude_mask(
+                    self.exclude_types,
+                    self.ntypes,
+                    self.sel_a,
```
Comment (njzjz) [inline deepmd/tf/descriptor/se_r.py:204]:
```suggestion
                    self.sel_r,
```

**Comment by njzjz** [comment]
> would deprecate the exclude type in the future.

What's the reason for that?

File: deepmd/pt/utils/env_mat_stat.py (hunk 1 above)
Comment (njzjz) [reply deepmd/pt/utils/env_mat_stat.py:151]: After I check the code, I think the zeros generated by the descriptor exclude types will still be passed to the neural network, so I think my implementation is correct.
The zeros generated by `pair_exclude_types` seems not used for stat. Need to confirm whether it is correct.

**Comment by wanghan-iapcm** [comment]
> > would deprecate the exclude type in the future.
> 
> What's the reason for that?

because not all atomic model has descriptor.  the dp atomic model's exclude pair is equivalent to the descriptors exclude pair.

**Comment by njzjz** [comment]
> > > would deprecate the exclude type in the future.
> 
> > 
> 
> > What's the reason for that?
> 
> 
> 
> because not all atomic model has descriptor.  the dp atomic model's exclude pair is equivalent to the descriptors exclude pair. 

An atomic model may not only contain one descriptor. Note that two descriptors in a DPRc model have different exclude pairs.

**Comment by wanghan-iapcm** [comment]
> > > > would deprecate the exclude type in the future.
> > 
> > 
> > > 
> > 
> > 
> > > What's the reason for that?
> > 
> > 
> > because not all atomic model has descriptor.  the dp atomic model's exclude pair is equivalent to the descriptors exclude pair.
> 
> An atomic model may not contain only one descriptor. Note that two descriptors in a DPRc model have different exclude pairs. So it may not be equivalent.

I see in multiple descriptor scenario they are not equivalent.

**Review by wanghan-iapcm — APPROVED** [review]

File: deepmd/pt/utils/env_mat_stat.py (hunk 1 above)
Comment (wanghan-iapcm) [reply deepmd/pt/utils/env_mat_stat.py:151]: As for descriptors, the pair exclusion is not used in stat.

**Review by iProzd — COMMENTED** [review]
I notice that the `excluded_types` only masks the input `env_mat`, while still keeps the total number of atoms in `excluded_types` (`sumn`). Another way is to mask `excluded_types` in both numerator (`sumr` etc.) and denominator (`sumn`) . Which one should we prefer? @wanghan-iapcm @njzjz

**Comment by wanghan-iapcm** [comment]
> I notice that the `excluded_types` only masks the input `env_mat`, while still keeps the total number of atoms in `excluded_types` (`sumn`). Another way is to mask `excluded_types` in both numerator (`sumr` etc.) and denominator (`sumn`) . Which one should we prefer? @wanghan-iapcm @njzjz

I would prefer mask both the numerator and the denominator, which gives consistent result as if the masked pairs do not exists.

**Comment by njzjz** [comment]
> > I notice that the `excluded_types` only masks the input `env_mat`, while still keeps the total number of atoms in `excluded_types` (`sumn`). Another way is to mask `excluded_types` in both numerator (`sumr` etc.) and denominator (`sumn`) . Which one should we prefer? @wanghan-iapcm @njzjz
> 
> I would prefer mask both the numerator and the denominator, which gives consistent result as if the masked pairs do not exists.

They are not equivalent. When masked pairs do not exist, zeros (meaning no neighbor) will be used for the stat.

The current masking behavior has some problems. The env mat is always masked to zero, but if `set_davg` is false, it should be masked to `davg`.

**Pull request merged**